# Verification of Answers to Problems 1

In [141]:
# Setup

import mujoco as mj
import mujoco.viewer
import time
import numpy as np
import sympy as sp

# Load the XML model

model = mj.MjModel.from_xml_path("penulum.xml")

data = mj.MjData(model)

joint1_name = "joint1"
joint2_name = "joint2"

joint1_id = model.joint(joint1_name).id
joint2_id = model.joint(joint2_name).id

world_frame_id = 0

frame1_name = "frame1"
frame1_id = model.body(name=frame1_name).id

frame2_name = "frame2"
frame2_id = model.body(name=frame2_name).id

end_effector_body_name="end-effector"
end_effector_body_id = model.body(name=end_effector_body_name).id

### Helper Functions

In [ ]:
def get_pose(body_id):
    #Extract the positions
    rot = data.xmat[body_id].reshape(3,3)
    pos = data.xpos[body_id]


    #Convert to transformation matrix
    transformation = np.eye(4) 
    transformation[:3, :3] = rot  
    transformation[:3, 3] = pos 
    return transformation

## Q1 Part A, B, and C

Using the values from question 1 part A: The joint angles **$\theta$** are given by 

$$
\theta = \begin{bmatrix} \theta_1 \\ \theta_2 \end{bmatrix} = \begin{bmatrix} 5\pi/6 \\ -3\pi/7 \end{bmatrix}
$$

Our poses we calculated with those angles is:
$$
{}^0T_C =
\begin{bmatrix}
-0.5   & -0.1927 & 0.8443 & 1.0754 \\
-0.8660& 0.1113  & -0.4875& 0.3028 \\
0      & -0.9749 & -0.2225& 1.0220 \\
0      & 0       & 0      & 1.0
\end{bmatrix}
$$
and
$$
{}^0T_2 =
\begin{bmatrix}
-0.1927 & 0.5 & 0.8443 & 0.4 \\
0.1113 & 0.8660 & -0.4875 & 0.6928 \\
-0.9749 & 0 & -0.2225 & 1.2 \\
0 & 0 & 0 & 1.0
\end{bmatrix}
$$

Now, using Mujoco we can plug in these **$\theta$** values and see that the simulated pose is the same. 



In [148]:
theta1_parta = 5 * np.pi / 6
theta2_parta = -3 * np.pi / 7

#set model to the given theta values
data.qpos[joint1_id] = theta1_parta
data.qpos[joint2_id] = theta2_parta

mj.mj_forward(model, data)

#Display
print("Resulting Pose of Frame c in respect to Frame 0:\n")
sp.pprint(sp.Matrix(np.round(get_pose(end_effector_body_id), 4)))

print("\nAs can be seen, this simulated result is the same as our calculated result.")


Resulting Pose of Frame c in respect to Frame 0:

⎡ -0.5   -0.1927  0.8443   1.0754⎤
⎢                                ⎥
⎢-0.866  0.1113   -0.4875  0.3028⎥
⎢                                ⎥
⎢ 0.0    -0.9749  -0.2225  1.022 ⎥
⎢                                ⎥
⎣ 0.0      0.0      0.0     1.0  ⎦

As can be seen, this simulated result is the same as our calculated result.


In [ ]:
#Display
print("Resulting Pose of Frame 2 in respect to Frame 0:\n")
sp.pprint(sp.Matrix(np.round(get_pose(frame2_id), 4)))

print("\nAs can be seen, this simulated result is the same as our calculated result.")


Resulting Pose of Frame 2 in respect to Frame 0:

⎡-0.1927   0.5   0.8443    0.4  ⎤
⎢                               ⎥
⎢0.1113   0.866  -0.4875  0.6928⎥
⎢                               ⎥
⎢-0.9749   0.0   -0.2225   1.2  ⎥
⎢                               ⎥
⎣  0.0     0.0     0.0     1.0  ⎦

As can be seen, this simulated result is the same as our calculated result.


Having seen that our simulated pose $ {}^0T_c $ and $ {}^0T_2 $ is the same as our calculated pose in question 1, the resulting quaternion and axis angle representations of the rotation matrixes are obviously the same as well.

## Q1 Part D

In Question 1d we are asked to find the values of theta corresponding to a given orientation of the camera frame. To verify this calculation we can plug in the values of theta that we found and see if the orientation matches what was given. 


The values of theta that result in that orientation we found are:

### Solution 1:
$$
(\theta_1, \theta_2) = (-0.7854, 5.4978)
$$

Resulted in this pose:
$$
\begin{bmatrix}
0.7071 & 0.5 & -0.5 & -0.9657 \\
0.7071 & -0.5 & 0.5 & -0.1657 \\
0 & -0.7071 & -0.7071 & 0.6343 \\
0 & 0 & 0 & 1
\end{bmatrix}
$$


### Solution 2:
$$
(\theta_1, \theta_2) = (2.356, 0.7854)
$$

Resulted in this pose:
$$
\begin{bmatrix}
-0.7071 & -0.5 & -0.5 & 0.1657 \\
-0.7071 & 0.5 & 0.5 & 0.9657 \\
0 & 0.7071 & -0.7071 & 0.6343 \\
0 & 0 & 0 & 1
\end{bmatrix}
$$

The resulting orientation is the z part of the rotation matrix:

$$
k = -\frac{1}{2} x_0 + \frac{1}{2} y_0 - \frac{\sqrt{2}}{2} z_0.
$$



In [ ]:
# Theta1 and theta2 solution optinos
solutions = [(-0.785398163397448, 5.49778714378214),
             (2.35619449019234, 0.785398163397448),]

a = 1
for solution in solutions:
    #set model to solution
    data.qpos[joint1_id] = solution[0]
    data.qpos[joint2_id] = solution[1]
    mj.mj_forward(model, data)
    
    print("Resulting Pose from Solution " + str(a) + ":\n")
    sp.pprint(sp.Matrix(np.round(get_pose(end_effector_body_id), 4)))
    print()

    k = data.xmat[end_effector_body_id].reshape(3,3)[:,2]
    print("Z Orientation extracted: \n")
    print(k)
    print("--------------")
    a+=1

print("As can be seen, this simulated result is the same as our calculated result.")

Resulting Pose from Solution 1:

⎡0.7071    0.5     -0.5    -0.9657⎤
⎢                                 ⎥
⎢0.7071   -0.5      0.5    -0.1657⎥
⎢                                 ⎥
⎢ 0.0    -0.7071  -0.7071  0.6343 ⎥
⎢                                 ⎥
⎣ 0.0      0.0      0.0      1.0  ⎦

Z Orientation extracted: 

[-0.5         0.5        -0.70710678]
--------------
Resulting Pose from Solution 2:

⎡-0.7071   -0.5    -0.5    0.1657⎤
⎢                                ⎥
⎢-0.7071   0.5      0.5    0.9657⎥
⎢                                ⎥
⎢  0.0    0.7071  -0.7071  0.6343⎥
⎢                                ⎥
⎣  0.0     0.0      0.0     1.0  ⎦

Z Orientation extracted: 

[-0.5         0.5        -0.70710678]
--------------
As can be seen, this simulated result is the same as our calculated result.


## Q1 Part E
In Question 1e we are asked to find the values of theta corresponding to a given position of the camera frame. To verify this calculation we can plug in the values of theta that we found and see if the orientation matches what was given. 


The values of theta that result in that position we found are:

### Solution 1:
We found
$$
(\theta_1, \theta_2) = (1.0727, 1.3464)
$$

that resulted in this pose:
$$
\begin{bmatrix}
-0.8785 & 0.1063 & 0.4658 & 1.0754 \\
0.4777 & 0.1955 & 0.8565 & 0.303 \\
0 & 0.9749 & -0.2225 & 1.022 \\
0 & 0 & 0 & 1
\end{bmatrix}
$$

### Solution 2:
We found
$$
(\theta_1, \theta_2) = (2.6181, 4.9368)
$$

That resulted in this pose:
$$
\begin{bmatrix}
-0.4999 & -0.1927 & 0.8444 & 1.0754 \\
-0.8661 & 0.1112 & -0.4873 & 0.303 \\
0 & -0.9749 & -0.2225 & 1.022 \\
0 & 0 & 0 & 1
\end{bmatrix}
$$
The resulting position is:
$$
p = 1.075x_0 + 0.303y_0 + 1.022z_0.
$$


In [169]:
solutions = [
(1.07272370972679, 1.34641832379787),
(2.61813591673788, -1.34641832379787)]

a=1
for solution in solutions:
    #set model to solution
    data.qpos[joint1_id] = solution[0]
    data.qpos[joint2_id] = solution[1]
    mj.mj_forward(model, data)
    print("Resulting Pose from Solution " + str(a) + ":\n")
    sp.pprint(sp.Matrix(np.round(get_pose(end_effector_body_id), 4)))
    print()

    p = data.xpos[end_effector_body_id]
    print("Z Orientation extracted: \n")
    print(p)
    print("--------------")
    a+=1


print("As can be seen, this simulated result is the same as our calculated result.")

Resulting Pose from Solution 1:

⎡-0.8785  0.1063  0.4658   1.0754⎤
⎢                                ⎥
⎢0.4777   0.1955  0.8565   0.303 ⎥
⎢                                ⎥
⎢  0.0    0.9749  -0.2225  1.022 ⎥
⎢                                ⎥
⎣  0.0     0.0      0.0     1.0  ⎦

Z Orientation extracted: 

[1.07541015 0.303      1.022     ]
--------------
Resulting Pose from Solution 2:

⎡-0.4999  -0.1927  0.8444   1.0754⎤
⎢                                 ⎥
⎢-0.8661  0.1112   -0.4873  0.303 ⎥
⎢                                 ⎥
⎢  0.0    -0.9749  -0.2225  1.022 ⎥
⎢                                 ⎥
⎣  0.0      0.0      0.0     1.0  ⎦

Z Orientation extracted: 

[1.07541015 0.303      1.022     ]
--------------
As can be seen, this simulated result is the same as our calculated result.


# Verification of Answers to Problem 2

#### Part A-Foreward kinematics

Using the values from question 1 part A: The joint angles **$\theta$** are given by 

$$
\theta = \begin{bmatrix} \theta_1 \\ \theta_2 \end{bmatrix} = \begin{bmatrix} 5\pi/6 \\ -3\pi/7 \end{bmatrix}
$$

The angular rates of change are also given:

$$
\dot{\boldsymbol{\theta}} = \begin{bmatrix} 1 & 2 \end{bmatrix}^{\top}
$$ 



The goal is to find the spatial and body twist, the spatial twist  found was:

$$
\left[\begin{matrix}-0.779942329745459\\1.77801674716505\\-0.779942329745459\\-2.0\\-0.974927912181824\\-0.222520933956314\end{matrix}\right]
$$


The body twist was: 
$$
\begin{bmatrix}
-2.07846096908265 \\
1.2 \\
0 \\
1.0 \\
1.73205080756888 \\
1.0
\end{bmatrix}
$$


In [147]:
# --- Reset the simulation and set the configuration ---
mj.mj_resetData(model, data)

# Set joint positions
data.qpos[joint1_id] = theta1_parta
data.qpos[joint2_id] = theta2_parta

# Set joint velocities: [1.0, 2.0]
data.qvel[joint1_id] = 1.0
data.qvel[joint2_id] = 2.0

# Compute the forward dynamics to update the simulation state
mj.mj_forward(model, data)

# --- Retrieve Sensor Data to Form the Body Twists ---

# Body sensors (end-effector's frame)
body_lin_sensor_id = model.sensor("body-linear-vel-c").id # 0
body_ang_sensor_id = model.sensor("body-ang-vel-c").id # 1

# Body twist: measured in the end-effector's (body) frame [linear; angular]
body_lin_vel = data.sensordata[0 : 3]
body_ang_vel = data.sensordata[3 : 6]
body_twist = np.concatenate([body_lin_vel, body_ang_vel])
print("\nResulting Body Twist (v,w):")
print(body_twist.round(4))
# Expected body twist:
# [ -0.779942329745459, 1.77801674716505, -0.779942329745459,
#   -2.0, -0.974927912181824, -0.222520933956314 ]


# find spatial twist using the body twist and an adjoint transformation
from spatialmath.base import tr2adjoint
spatial_twist = tr2adjoint(np.array(get_pose(end_effector_body_id))) @ body_twist

print("\nResulting Spatial Twist (v,w):")
print(spatial_twist.round(4))
# Expected spatial twist:
# [ -2.07846096908265, 1.2, 0, 1.0, 1.73205080756888, 1.0 ]

print("\nAs can be seen, the simulated twist is the same as the calculated twist.")



Resulting Body Twist (v,w):
[-0.7799  1.778  -0.7799 -2.     -0.9749 -0.2225]

Resulting Spatial Twist (v,w):
[-2.0785  1.2     0.      1.      1.7321  1.    ]

As can be seen, the simulated twist is the same as the calculated twist.


## Inverse Velocity Kinematics


there is still lots to do. verify everything